docs: Make use of new curl example system

hackerkid opened the pull request:
Related issue #12878

hackerkid commented:
@timabbott While migrating I noticed two issues.

* `generate_code_example` appends `path` params as `query` params in curl example. So I have to manually exclude them using the `exclude` argument. I can probably update the code to take care of this case automatically.

  ```
  curl -sSX GET -G http://localhost:9991/api/v1/messages/{message_id} \
    -u BOT_EMAIL_ADDRESS:BOT_API_KEY \
    -d 'message_id=42'
  ```
  ```
  {generate_code_example(curl, exclude= ["message_id"])|/messages/{message_id}:get|example}
  ```
  http://zulip.zulipdev.com:9991/api/get-raw-message

* In some docs there are two curl examples. For example http://zulip.zulipdev.com:9991/api/send-message. `generate_code_example` can only generate one. What should be done in this case?

hackerkid commented:
I added some commits to take care of first and second issue. Another issue I noticed is that `test_generated_curl_examples_for_success` just run the curl command and see if the response is success. That wont work in most cases since the example values present in zulip.yaml might not present in the test database at all and can  result in 400 status code. Also not sure whether we have to run all the generated curl examples since there is no reason it wont work with proper values as long as the openapi specs are correct.

timabbott commented:
The first part of this branch is great; I added a quick test for the uncovered code that made this fail CI and merged all but the last three commits.

> That wont work in most cases since the example values present in zulip.yaml might not present in the test database at all and can result in 400 status code. 

Ideally, we'd figure out a way to make the example data work here.  Might require something clever to figure out what IDs to use dynamically, similar to how our Python API example tests have some wrapping code that fetches objects from the database to get the IDs, and then pass that into the code running the test.  

> Also not sure whether we have to run all the generated curl examples since there is no reason it wont work with proper values as long as the openapi specs are correct.

I think it's helpful to have an end-to-end test; this helps us verify the specs are correct (as we've seen!).

timabbott commented:
Awesome, thanks for working on this @hackerkid!  Unfortunately, there are several bugs in the replacement documentation, basically due to the OpenAPI data being wrong.  For example, with `/update-subscription-properties.md`, if you substitute in the email/API key from `manage.py print_initial_password iago@zulip.com` and then deal with the `stream_id` values being wrong, you'll find that `f00` is forbidden by our validators; a proper color should be e.g. `#f00f00`.  This is the reason I wanted the curl examples to actually be tested by running them.  

I think it's probably reasonable for you to just hand-test all of the examples using the approach I just described and include manual fixes for the bugs like that so we can merge this, since it's definitely an improvement to migrate systems.  

I think if we do want to make automated tests work, we just need a way to substitute example values.  I think we could have a data structure like this:
```
{'/users/me/subscriptions/properties:post|@@ -124,6 +124,7 @@ def get_openapi_param_example_value_as_string(endpoint: str, method: str, param:
                                               curl_argument: bool=False) -> str:
     param_type = param["schema"]["type"]
     param_name = param["name"]
+    if settings.RUNNING_OPENAPI_CURL_TESTS:
+        patched_param = patch_param_example_ids(endpoint, method, param)
     if param_type in ["object", "array"]:
         example_value = param.get("example", None)
         if not example_value:
```

And then that patching method would basically just have a bit of per-example code to e.g. replace the `stream_id`s or `message_id`s in the example data structures from the OpenAPI data by looking up actual things in the database.  

And then we're write this function `patch_param_example_ids`

But it probably makes sense to work on that once we have this merged with manual testing, just because the manual testing will probably take only an hour and it'd let us get this big PR integrated.

hackerkid commented:
@timabbott I have converted most of the endpoints. One remaining are 

* `update-stream` - Issue with how some string params are quoted.
* `update-notification-settings` - Issue with how some string params are quoted.
* `send-message` - The doc has 2 Curl examples with diffrent values for each params. This is the only case like this. I can probably use curl example generator to show one param and keep the other curl example just as it is.
* `upload-file` - Does not have any example in doc. Can be generated by making curl example generator handle file upload.
* `upload-custom-emoji` - Has an example but need modification in curl example generator to make file upload work.

timabbott commented:
Great; can you comment on whether you've successfully done manual testing for all the converted endpoints?

hackerkid commented:
@timabbott Yup. Did test all the endpoints I converted manually. I have also added some commits in between to fix some mismatch I saw between the docs and actual behavior of some endpoints.

timabbott commented:
I just did a review of these, and this is great, thanks @hackerkid (and thanks @Hypro999 for building the system that makes this possible).

> update-stream - Issue with how some string params are quoted.
> update-notification-settings - Issue with how some string params are quoted.

We should figure out what the issue is; can you post details?  It might be something @Hypro999 or I understand how to fix easily.

> send-message - The doc has 2 Curl examples with different values for each params. This is the only case like this. I can probably use curl example generator to show one param and keep the other curl example just as it is.

That's probably is reasonable as a near-term plan.  If we choose to make the example parameter values mockable as I described above, I could imagine us doing something with that.

> upload-file - Does not have any example in doc. Can be generated by making curl example generator handle file upload.
> upload-custom-emoji - Has an example but need modification in curl example generator to make file upload work.

Makes sense.  Let's do these if it's easy.

In any case we should probably finish off what we're doing on this front now and then open a follow-up issue that just describes the remainder.

Huge thanks for doing this migration @hackerkid!  I think it'll make a big difference in our ability to keep this documentation current.

hackerkid commented:
> update-stream - Issue with how some string params are quoted.
update-notification-settings - Issue with how some string params are quoted.

So the issue is that these endpoints expect valid JSON strings (check_string; enclosed in double quotes and special characters escaped with backslashes). curl example generator does not make strings JSON valid currently.

```
curl -sSX PATCH http://zulip.zulipdev.com:9991/api/v1/streams/4 \ 
    -u iago@zulip.com:y2cQK33rJUDxJ9b0xxFqSfhHnPxSZCmF \
    -d 'description=This stream is related to football dicsussions.' \ 
    -d 'new_name="Italy3"' \
    -d 'is_private=true' \
    -d 'is_announcement_only=true' \
    -d 'history_public_to_subscribers=true'
{"result":"error","msg":"Argument \"description\" is not valid JSON.","code":"BAD_REQUEST"}
```

Hypro999 commented:
@hackerkid can you give an example of what a valid curl example would look like for this case (just to make things a little clearer)? I think that the best approach might be to make this an option that we support with the kwargs parameter that we create in `render_curl_example` in `zerver/lib/openapi/api_code_examples.py`. So we could allow a string like `{generate_code_example(curl, jsonify=["param3"])|/messages:get|example}` or something along these lines... Similar to how you have added the include option alongside the exclude option.

hackerkid commented:
>  can you give an example of what a valid curl example would look like for this case (just to make things a little clearer)?

Sure.

```bash
curl -sSX PATCH http://zulip.zulipdev.com:9991/api/v1/streams/4 \
    -u iago@zulip.com:y2cQK33rJUDxJ9b0xxFqSfhHnPxSZCmF \
    -d 'description="This stream is related to football dicsussions."' \
    -d 'new_name="Italy3"' \    
    -d 'is_private=true' \
    -d 'is_announcement_only=true' \
    -d 'history_public_to_subscribers=true'
{"result":"success","msg":""}
```

> I think that the best approach might be to make this an option that we support with the kwargs parameter that we create in render_curl_example in zerver/lib/openapi/api_code_examples.py. So we could allow a string like {generate_code_example(curl, jsonify=["param3"])|/messages:get|example} or something along these lines.

hmm. I think this info should be stored in OpenAPI specs instead of docs so that the person/machine who is reading the OpenAPI specs also knows to do the right thing. One way to do this is by making use of the "format" keyword.

```diff
         description: The new description for the stream.
         schema:
           type: string
+          format: json
         example: "This stream is related to football dicsussions."
```

```diff
         example_value = param.get("example", DEFAULT_EXAMPLE[param["schema"]["type"]])
         if type(example_value) == bool:
             example_value = str(example_value).lower()
+        if param["schema"].get("format", "") == "json":
+            example_value = json.dumps(example_value)
         if curl_argument:
             return "    -d '{}={}'".format(param_name, example_value)
```

What do you folks think of this approach?

timabbott commented:
Yeah, I think the `format` keyword is the right answer here.   Nice find.
